accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+        and `V`.
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is an `ifElement` with two `element`s, `p1` and `p2`:
+
+    The condition is inferred with a context type of `bool`.
+
+    It is a compile error if `p1` has an inferred set element type and `p2` does
+    not, or if `p2` has an inferred set element type and `p1` does not. *In
+    other words, you can't spread a map on one branch and a set on the other.
+    Since `dynamic` provides both set and map key/value types, a `dynamic` in
+    either branch does not run into this case.*
+
+    *   If the inferred set element type of `p1` is `S1` and the inferred set
+        element type of `p2` is `S2` then the inferred set element type of
+        `element` is the upper bound of `S1` and `S2`.
+
+    *   If the inferred key type of `p1` is `K1` and the inferred key type of
+        `p1` is `V1` and the inferred key type of `p2` is `K2` and the inferred
+        key type of `p2` is `V2` then the inferred key type of `element` is the
+        upper bound of `K1` and `K2` and the inferred value type is the upper
+        bound of `V1` and `V2`.
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is a `forElement` with `element` `p1` then:
+
+    Inference for the iterated expression and the controlling variable is done
+    as for the corresponding `for` or `await for` statement.
+
+    *   If the inferred set element type of `p1` is `S` then the inferred set
+        element type of `element` is `S`.
+
+    *   If the inferred key type of `p1` is `K` and the inferred key type of
+        `p1` is `V` then the inferred key and value types of `element` are `K`
+        and `V`.
+
+    *In other words, inference flows upwards from the body element. Note that
+    both of the above cases can validly apply because of `dynamic` spreads.*
+
+Finally, we define inference on a `setOrMapLiteral` `collection` as follows:
+
+*   If `collection` is a set literal, then the downwards context for inference
+    of the elements of `collection` is `Set<P>` where `P` may be `?` if
+    downwards inference does not constrain the type of `collection`.
+
+    *   If `P` is `?` then the static type of `collection` is `Set<T>` where `T`
+        is the upper bound of the inferred set element types of the elements.
+
+    *   Otherwise, the static type of `collection` is `Set<T>` where `T` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a key/value type here
+    given that downwards context.*
+
+*   If `collection` is a map literal and downwards inference has statically
+    known type `Map<K, V>` then the downwards context for the elements of
+    `collection` is `Map<K, V>`.
+
+*   If `collection` is a map literal then the downwards context for the elements
```

Comment:
Done. I'm never sure the best way to express a series of disjoint options in pseudo-code. I don't want to use "Else, if" since that implies the order matters, which requires a closer reading and doesn't apply here. So I used a bullet list instead of a number list and a series of bare "ifs". But, yeah, I think "Otherwise" is nicer for the last.